lib/puppet/configurer.rb

```diff
@@ -207,7 +207,7 @@ def run_internal(options)
       if catalog = prepare_and_retrieve_catalog_from_cache
         options[:catalog] = catalog
         @cached_catalog_status = 'explicitly_requested'
-        @environment = catalog.environment
+        @environment = catalog.environment unless Puppet[:strict_environment_mode]
```

Comment:
I'm wondering if `strict_environment_mode` should apply when using a cached catalog? I'm thinking the least surprising thing would be to always use the environment in the cached catalog.

Otherwise, I think we need to take into account whether the user specified an environment on the command line/settings, and really did intend to only use the cached catalog from a particular environment. That would be a lot easier if the agent didn't default to `production` so we could differentiate between `nil` (has no agent environment) vs `agent-specified environment`.

Thoughts @MikaelSmith?

Adding a notice message when switching environments sounds good.